From 612278e2de5989c69d86c860566c22ca047fcb25 Mon Sep 17 00:00:00 2001
From: Carlos Valiente <carlos@pepelabs.net>
Date: Mon, 28 Jun 2010 19:06:09 +0100
Subject: [PATCH] Speed up `reactor.spawnProcess()`.

On platforms that list open file descriptors under '/proc/<pid>', this
patch makes `reactor.spawnProcess()` close open file descriptors only,
instead of the whole range [0, <value of ulimit -n>].
---
 twisted/internet/process.py |   23 +++++++++++++++--------
 1 files changed, 15 insertions(+), 8 deletions(-)

diff --git a/twisted/internet/process.py b/twisted/internet/process.py
index c5f1ac3..88040a0 100644
--- a/twisted/internet/process.py
+++ b/twisted/internet/process.py
@@ -601,7 +601,8 @@ class Process(_BaseProcess):
         This is accomplished in two steps::
 
             1. close all file descriptors that aren't values of fdmap.  This
-               means 0 .. maxfds.
+               means 0 .. maxfds (or just the open fds within that range, if
+               the platform supports '/proc/<pid>/fd').
 
             2. for each childFD::
 
@@ -626,15 +627,21 @@ class Process(_BaseProcess):
 
         destList = fdmap.values()
         try:
-            import resource
-            maxfds = resource.getrlimit(resource.RLIMIT_NOFILE)[1] + 1
-            # OS-X reports 9223372036854775808. That's a lot of fds to close
-            if maxfds > 1024:
-                maxfds = 1024
+            proc_fd = "/proc/%d/fd" % os.getpid()
+            openfds = tuple(int(fd) for fd in os.listdir(proc_fd))
         except:
-            maxfds = 256
+            try:
+                import resource
+                maxfds = resource.getrlimit(resource.RLIMIT_NOFILE)[1] + 1
+                # OS-X reports 9223372036854775808. That's a lot of fds
+                # to close
+                if maxfds > 1024:
+                    maxfds = 1024
+            except:
+                maxfds = 256
+            openfds = xrange(maxfds)
 
-        for fd in xrange(maxfds):
+        for fd in openfds:
             if fd in destList:
                 continue
             if debug and fd == errfd.fileno():
-- 
1.6.3.3

